File: game/declaration.go
package game

type Declaration interface {
	Points() int
}

type PreHandDeclarationType int

type PreHandDeclaration struct {
	Type        PreHandDeclarationType
	HighestCard Card
}

type Belote struct{}

const (
	DeclarationNone PreHandDeclarationType = iota
	Tierce
	Quarte
	Quinte
	Carre
	NinesCarre
	JacksCarre
)

var declarationPoints = map[PreHandDeclarationType]int{
	DeclarationNone: 0,
	Tierce:          20,
	Quarte:          50,
	Quinte:          100,
	Carre:           100,
	NinesCarre:      150,
	JacksCarre:      200,
}

func (b Belote) Points() int {
	return 20
}

func (d PreHandDeclaration) Points() int {
	return declarationPoints[d.Type]
}

func CompareDeclarations(d1, d2 PreHandDeclaration, trump Suit) int {
	if d1.Type != d2.Type {
		return int(d1.Type) - int(d2.Type)
	}

	if d1.HighestCard.Suit == trump && d2.HighestCard.Suit != trump {
		return 1
	} else if d1.HighestCard.Suit != trump && d2.HighestCard.Suit == trump {
		return -1
	}

	return d1.HighestCard.Rank.NaturalOrder() - d2.HighestCard.Rank.NaturalOrder()
}

File: client/lib/src/state/declaration.ts
import { IsEnum, IsInt, IsOptional, ValidateNested } from 'class-validator'
import { Type } from 'class-transformer'
import { Card } from '../card'

export enum DeclarationType {
	Tierce = 'Tierce',
	Quarte = 'Quarte',
	Quinte = 'Quinte',
	Carre = 'Carre',
	NinesCarre = 'NinesCarre',
	JacksCarre = 'JacksCarre',
	Belote = 'Belote',
}

export function getDeclarationTypeName(type: DeclarationType): string {
	switch (type) {
		case DeclarationType.Tierce:
			return 'Tierce'
		case DeclarationType.Quarte:
			return 'Quarte'
		case DeclarationType.Quinte:
			return 'Quinte'
		case DeclarationType.Carre:
			return 'Carré'
		case DeclarationType.NinesCarre:
			return 'Carré'
		case DeclarationType.JacksCarre:
			return 'Carré'
		case DeclarationType.Belote:
			return 'Belote'
	}
}

export class Declaration {
	@IsEnum(DeclarationType)
	type: DeclarationType

	@Type(() => Card)
	@IsOptional()
	@ValidateNested()
	highestCard?: Card

	@IsInt()
	points: number

	constructor(type: DeclarationType, points: number, highestCard?: Card) {
		this.type = type
		this.points = points
		this.highestCard = highestCard
	}
}
